from django import forms
from django.core.exceptions import ValidationError
from django.contrib.auth import password_validation

from .models import User


class UserCreationForm(forms.ModelForm):
    password1 = forms.CharField(strip=False,)
    password2 = forms.CharField(strip=False,)

    class Meta:
        model = User
        fields = ("username",)

    def clean_password2(self):
        password1 = self.cleaned_data.get("password1")
        password2 = self.cleaned_data.get("password2")
        if password1 and password2 and password1 != password2:
            raise ValidationError('The two password fields didnâ€™t match.')
        return password2

    def _post_clean(self):
        super()._post_clean()
        password = self.cleaned_data.get('password2')
        if password:
            try:
                password_validation.validate_password(password, self.instance)
            except ValidationError as error:
                self.add_error('password2', error)

    def save(self, commit=True):
        user = super().save(commit=False)
        user.set_password(self.cleaned_data["password1"])
        if commit:
            user.save()
        return user
